1653 Option to rebind your inputs

FuJa0815 opened the pull request:
Closes #1653 
Add option to rebind your inputs.

hhyyrylainen commented on src/engine/input/InputAction.cs:
```diff
@@ -0,0 +1,102 @@
+    {
+        {
+            "Movement", ToReadonly(
+                new InputAction("g_move_forward",   "Move forward"),
```
May I suggest that this information would be loaded from JSON (in simulation parameters)? That way you can edit this info without touching the code (which would make sense as you can configure new inputs in the godot project, which also doesn't need programming).

FuJa0815 replied:
Good idea

FuJa0815 commented:
Ok, I really need some help with godot and creating GUI elements at runtime.
I have created a new tab in the Options menu.
In the new tabs' content I added a VBoxContainer with the InputGroupList script added to it.
The InputGroupList script loads the json from the file, loads the scenes, instanciates the scenes and adds new Childs.
The json loading works, but the problem is: the new Children are not visible.

For testing I have added this line
https://github.com/Revolutionary-Games/Thrive/blob/236a91e11b74727394d50adf78ba451a20b471f1/src/engine/input/key-mapping/InputGroupList.cs#L9-L14
but the label also does not get renderd.

I have googled a lot, but didn't find an answer.
Does anyone know why my new elements don't get shown on the screen.
If I manually add a Label to the InputGroupList-VBoxContainer it gets rendered just fine.

hhyyrylainen commented:
I don't know what's up with the first part of the format script, but the second change it makes is it removed a newline from the end of the file. I strongly recommend you setup things so that you can run it locally.

Regarding the invisible children, perhaps you need to set the vbox to have the vertical expand size flag?
Btw when running the game from Godot editor, you can inspect the remote scene tree and edit it in realtime to see if the children are properly added and the element sizes end up correct.

FuJa0815 commented:
Thank you!
Everything worked fine, but one tiny line of code:
https://github.com/Revolutionary-Games/Thrive/blob/236a91e11b74727394d50adf78ba451a20b471f1/src/general/OptionsMenu.cs#L449-L456
yeah, I feel pretty dumb now

FuJa0815 commented:
Still have no idea what JSONLint is complaining about.

Everything should be implemented now.

hhyyrylainen commented:
So I opened the given diff from circleci in emacs to see what's up, and there is a "ZERO WIDTH NO-BREAK SPACE" character that jsonlint wants to be removed:
![Kuvakaappaus - 2020-10-29 08-49-55](https://user-images.githubusercontent.com/3796411/97535310-1fa31d00-19c4-11eb-8aff-30d5e145d5ba.png)

FuJa0815 commented:
I could not see the invisible char in a hex editor, but I changed CR LF to LF, which seemed to work

hhyyrylainen commented on src/engine/Settings.cs:
```diff
@@ -174,6 +176,14 @@ private Settings()
+    ///   Key = InputMap name.
+    ///   Value = The events associated with the control.
+    /// </summary>
+    public SettingValue<Dictionary<string, List<InputEventWithModifiers>>> CurrentControlScheme { get; set; } =
```
I don't think the control scheme needs saving, even if we add one. Control schemes should just be a list of presets that override the currently configured keys.

hhyyrylainen replied:
After reading everything, turns out these are the current controls?
I think this should be renamed and if `Dictionary<string, List<InputEventWithModifiers>>` is changed to be a class, this'll be cleaned up pretty nicely.

hhyyrylainen commented on src/engine/Settings.cs:
```diff
@@ -337,6 +348,25 @@ public void ApplySoundSettings()
+            // Add the new input events
+            foreach (var inputEvent in action.Value)
+            {
+                InputMap.ActionAddEvent(action.Key, inputEvent);
```
Does this work if there is an action specified in the input json that doesn't exist in Godot?
I think it might make sense to remove the inputs configured in godot editor, as the json needs to be also updated, right?

FuJa0815 replied:
no!
The godot inputs is where the game gets the default inputs from

hhyyrylainen replied:
I see. I don't remember seeing duplicate input definitions. I made this comment before reading and digesting all of the changes in this PR.
It's good to keep things as is (default keys configured just once in Godot editor).

FuJa0815 replied:
You can still just define new inputs in the godot editor and everything works, but you cannot rebind them in the options.
If you want the user to be able to rebind the new input you need to insert it to the json

hhyyrylainen replied:
That's exactly how it should work.
Though, I have one more concern. Does the system correctly handle new inputs being added later? That should work so that the new inputs are added to the active inputs even if the player has edited other inputs.
So the logic would be: on startup add any input keys that don't exist in the current input scheme to the scheme to ensure that things work fine when new actions are added.

FuJa0815 replied:
I think runtime additions to the InputMap should work fine.
I can't think of any reason why runtime additions to the InputMap should fail.
Ill test that later

hhyyrylainen replied:
I'm not talking about runtime additions. I'm talking about a potential bug:
- Player plays the game
- Edits the keys a bit
- A new release is released, it adds a new input action name to the game
- The user starts the game, and now they can't use that new input or the input binding doesn't show it

To work against this, on start the game should go through the InputMap and add any inputs to the current control scheme that are missing.

FuJa0815 replied:
I have tested it and the removal and addition of inputs works

hhyyrylainen replied:
That's good. It's really difficult to see from the quite large amount of code if that is handled correctly or not.

hhyyrylainen commented on src/engine/Settings.cs:
```diff
@@ -383,7 +413,11 @@ private static Settings LoadSettings()
-                return JsonConvert.DeserializeObject<Settings>(text);
+                var result = JsonConvert.DeserializeObject<Settings>(text, new InputEventWithModifiersConverter());
+                if (result == null)
+                    throw new Exception();
```
Exception should really have a message

hhyyrylainen commented on src/engine/input/key-mapping/InputActionItem.cs:
```diff
@@ -0,0 +1,100 @@
+using Godot;
+using Newtonsoft.Json;
+
+public class InputActionItem : VBoxContainer
```
Add documentation for what this class is for

hhyyrylainen commented on src/engine/input/key-mapping/InputActionItem.cs:
```diff
@@ -0,0 +1,100 @@
+    private HBoxContainer inputEventsContainer;
+    private Button addInputEvent;
+
+    [JsonProperty]
```
As this is a GUI element in the options menu, this doesn't need saving, right?
Why did you add these attributes here?

hhyyrylainen commented on src/engine/input/key-mapping/InputEventItem.cs:
```diff
@@ -0,0 +1,138 @@
+﻿using System;
+using Godot;
+
+public class InputEventItem : Node
```
documentation

hhyyrylainen commented on src/engine/input/key-mapping/InputEventItem.cs:
```diff
@@ -0,0 +1,138 @@
+    [Export]
+    public NodePath ButtonPath;
+
+    internal InputActionItem AssociatedAction;
+
+    internal bool JustAdded;
```
Why are these internal? If these need to be externally set from somewhere else these should be properties

hhyyrylainen commented on src/engine/input/key-mapping/InputEventItem.cs:
```diff
@@ -0,0 +1,138 @@
+        AssociatedAction.Inputs.Remove(this);
+    }
+
+    public override void _Input(InputEvent @event)
```
This seems to perform key reassinging? Please add doc on this method about that.

hhyyrylainen commented on src/engine/input/key-mapping/InputEventItem.cs:
```diff
@@ -0,0 +1,138 @@
+
+    private bool waitingForInput;
+
+    public InputEventWithModifiers AssociatedEvent { get; set; }
```
I don't feel particularly good about storing the Godot input events? Which inherits from Resource, so this is highly dubious to me. If this information truly needs to be stored, we need a custom class that grabs the needed info from an InputEventWithModifiers.

FuJa0815 replied:
storing the InputEventWithModifiers makes the modification of the InputMap much easier.
To save just the needed information I have created a jsonConverter

hhyyrylainen replied:
Reading this: `<para>Unlike <see cref="T:Godot.Object" />s, References keep an internal reference counter so that they are automatically released when no longer in use, and only then. References therefore do not need to be freed manually with <see cref="M:Godot.Object.Free" />.` bit of documentation on Reference, makes it a bit more stomachable for me to use InputEventWithModifiers in this manner.

FuJa0815 replied:
It states that i do not need to free them manually, but what if I do it anyways?
Shouldn't some calls to Free resolve any memory leaks without any major code reworks?

hhyyrylainen replied:
The Godot input tutorials don't call Free on the input events passed to scripts. I'm not sure if it causes any problems to call Free manually. At least with deleting Nodes you run the risk of getting object disposed exceptions when you don't want to get them.

FuJa0815 replied:
The Free method exists, but it prints
`Can't 'free' a reference.`
To the console.
Calling dispose() gives me ObjectDisposedExceptions...

hhyyrylainen replied:
Then you should follow the documentation and not call free.
The disposed exceptions indicate to me that you are disposing the thing at inopportune times.
This is why I'd really prefer our own class that behaves exactly as needed for the input key binding tracking, even if that means some extra code to convert between that and the godot types.

hhyyrylainen commented on src/engine/input/key-mapping/InputEventItem.cs:
```diff
@@ -0,0 +1,138 @@
+
+        var old = AssociatedEvent;
+        AssociatedEvent = (InputEventWithModifiers)@event;
+        var conflict = InputGroupList.Instance.Conflicts(this);
```
I think that `this` shouldn't be passed to get conflicts. Instead whatever info Conflicts wants from this should be passed in.

FuJa0815 replied:
this is outdated and the InputEventItem is now required in the Conflicts method.

hhyyrylainen replied:
I still think Conflicts should be coded so that it doesn't rely on this specific InputEventItem class.

hhyyrylainen commented on src/engine/input/key-mapping/InputEventItem.cs:
```diff
@@ -0,0 +1,138 @@
+            }
+        }
+
+        var old = AssociatedEvent;
+        AssociatedEvent = (InputEventWithModifiers)@event;
+        var conflict = InputGroupList.Instance.Conflicts(this);
+        if (conflict != null)
+        {
+            InputGroupList.Instance.ShowDialog(this, conflict, AssociatedEvent);
+            AssociatedEvent = old;
+            return;
+        }
+
+        waitingForInput = false;
+        JustAdded = false;
+        AssociatedAction.AssociatedGroup.GetParent<InputGroupList>().ControlSchemeChanged();
+
+        UpdateButtonText();
```
I think this area needs some comments / splitting into separate methods.

hhyyrylainen commented on src/engine/input/key-mapping/InputEventItem.cs:
```diff
@@ -0,0 +1,138 @@
+        UpdateButtonText();
+    }
+
+    public override void _Ready()
```
Ready should be before other godot callbacks.

hhyyrylainen commented on src/engine/input/key-mapping/InputEventItem.cs:
```diff
@@ -0,0 +1,138 @@
+            UpdateButtonText();
+    }
+
+    public override bool Equals(object obj)
```
Are all of these custom Equals and GetHashCode methods needed?

FuJa0815 replied:
they are required by the Conflicts method

hhyyrylainen commented on src/engine/input/key-mapping/InputEventItem.cs:
```diff
@@ -0,0 +1,138 @@
+                    text += "Shift+";
+                text += iemb.ButtonIndex switch
+                {
+                    1 => "Left mouse",
```
Would it be possible to get these strings from godot?

FuJa0815 replied:
no, because the AsText and the ToString methods of the InputMouseButton return a technical string

hhyyrylainen commented on src/engine/input/key-mapping/InputEventWithModifiersConverter.cs:
```diff
@@ -0,0 +1,71 @@
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+public class InputEventWithModifiersConverter : JsonConverter<InputEventWithModifiers>
```
I'm a bit worried that as this creates Godot.Reference inheriting objects, this can permanently leak objects.
See my previous comment about using a custom class instead than InputEventWithModifiers.

hhyyrylainen replied:
The Node based deserializer in many cases will register the deserialized objects to be freed later.

FuJa0815 replied:
But I do require a InputEventWithModifiers at some point for passing it to the InputMap class.
I thought that working with one class only would lead to better code, but if this really produces memory leaks I will change this

hhyyrylainen replied:
On reviewing the Reference class again, it kinda seems it doesn't. But still that class is in Godot and implemented in C++, meaning that marshalling etc. call overhead from C# is incurred whenever it is used.
If a separate class is created, it's obvious that it needs a method to create an equivalent InputEventWithModifiers for use elsewhere.

hhyyrylainen commented on src/engine/input/key-mapping/InputGroupItem.cs:
```diff
@@ -0,0 +1,40 @@
+using Godot;
+using Newtonsoft.Json;
+
+public class InputGroupItem : VBoxContainer
```
documentation

hhyyrylainen commented on src/engine/input/key-mapping/InputGroupItem.cs:
```diff
@@ -0,0 +1,40 @@
+public class InputGroupItem : VBoxContainer
+{
+    [Export]
+    [JsonIgnore]
```
Are json properties in this needed?

FuJa0815 replied:
those JsonIgnore properties were useless, because those were not props.
The JsonPropertyAttributes are redundant, but clarify where the data comes from.
Should I remove the unnecessary JsonPropertyAttributes?

hhyyrylainen replied:
They should be removed. I'm very confused why you are using JSON serialization on the GUI nodes? And in the options menu of all places, that just seems *wrong* to me.

hhyyrylainen commented on src/engine/input/key-mapping/InputGroupItem.cs:
```diff
@@ -0,0 +1,40 @@
+
+        inputGroupHeader.Text = GroupName;
+
+        // Add the actions to the DOM
```
DOM is a browser specific term

hhyyrylainen commented on src/engine/input/key-mapping/InputGroupItem.cs:
```diff
@@ -0,0 +1,40 @@
+    public string GroupName { get; set; }
+
+    [JsonProperty]
+    public IList<InputActionItem> Actions { get; set; }
```
List should be used rather than IList, unless there's a good reason?

FuJa0815 replied:
There isn't really a reason.
Generally, you should create variables of the interface, not the implementation, but List is a very common class, so you're right.

hhyyrylainen commented on src/engine/input/key-mapping/InputGroupList.cs:
```diff
@@ -0,0 +1,180 @@
+using Newtonsoft.Json;
+
+/// <summary>
+///   Used by OptionsMenu>Inputs>InputGroupContainer
```
Could you also mention very quickly what this class does?

hhyyrylainen commented on src/engine/input/key-mapping/InputGroupList.cs:
```diff
@@ -0,0 +1,180 @@
+    [Export]
+    public NodePath ConflictDialogPath;
+
+    internal static readonly PackedScene InputEventItemScene =
```
Scene should be generally loaded in _Ready

hhyyrylainen commented on src/engine/input/key-mapping/InputGroupList.cs:
```diff
@@ -0,0 +1,180 @@
+    internal static readonly PackedScene InputEventItemScene =
+        GD.Load<PackedScene>("res://src/engine/input/key-mapping/InputEventItem.tscn");
+
+    private static readonly Dictionary<string, List<InputEventWithModifiers>> DefaultControlScheme =
+        GetCurrentlyAppliedControlScheme();
```
Readonly properties are only assigned when first read, so this might not actually get the default control scheme if read after it is changed?

hhyyrylainen commented on src/engine/input/key-mapping/InputGroupList.cs:
```diff
@@ -0,0 +1,180 @@
+
+    public event ControlSchemeChangedDelegate OnControlSchemeChanged;
+
+    public static InputGroupList Instance { get; private set; }
```
Instead of making this a static property, could it be made so that the child objects are passed callbacks etc. so that they don't need to access this like this?
Also this reference doesn't seem to be unset anywhere so the garbage collector can't dispose of this.

FuJa0815 replied:
It is now used by the pause menu, which means that I require this property.
The reference does not get unset per se, but it gets overritten when there is a new OptionsMenu getting loaded.
I think I have found another memory issue with circular references, which I will resolve.

hhyyrylainen commented on src/engine/input/key-mapping/InputGroupList.cs:
```diff
@@ -0,0 +1,180 @@
+    public event ControlSchemeChangedDelegate OnControlSchemeChanged;
+
+    public static InputGroupList Instance { get; private set; }
+    public IEnumerable<InputGroupItem> AllGroupItems => GetChildren().OfType<InputGroupItem>();
```
If children are needed to be accessed often they should be kept in a separate list when adding them.

hhyyrylainen commented on src/engine/input/key-mapping/InputGroupList.cs:
```diff
@@ -0,0 +1,180 @@
+
+    public static Dictionary<string, List<InputEventWithModifiers>> GetCurrentlyAppliedControlScheme()
+    {
+        return InputMap.GetActions().OfType<string>()
+            .ToDictionary(p => p,
+                p => InputMap.GetActionList(p).OfType<InputEventWithModifiers>()
+                    .ToList());
+    }
```
This is really complex? Why is it so complex to know the current controls?

FuJa0815 replied:
I have not found a better solution to get the current godot inputs

hhyyrylainen replied:
Then could you write a wrapper class or something?
Reflecting back, I think that maybe you should rework your code to be way less tightly coupled to Godot's InputMap by introducing wrapper classes that are used in place of the InputMap, until it is time to apply the final results.

hhyyrylainen commented on src/engine/input/key-mapping/InputGroupList.cs:
```diff
@@ -0,0 +1,180 @@
+                    .ToList());
+    }
+
+    public Dictionary<string, List<InputEventWithModifiers>> GetCurrentlyPendingControlScheme()
```
`Dictionary<string, List<InputEventWithModifiers>>` is used a ton, it would be nice to wrap that in a class like `ThriveInputMap` (couldn't think of a better name).

hhyyrylainen commented on src/engine/input/key-mapping/InputGroupList.cs:
```diff
@@ -0,0 +1,180 @@
+        conflictDialog = GetNode<ConfirmationDialog>(ConflictDialogPath);
+    }
+
+    public void ShowDialog(InputEventItem caller, InputEventItem conflict, InputEventWithModifiers newEvent)
```
ShowDialog sounds like a too general name

hhyyrylainen commented on src/engine/input/key-mapping/InputGroupList.cs:
```diff
@@ -0,0 +1,180 @@
+    /// </summary>
+    /// <param name="item">The event with the new value</param>
+    /// <returns>The collision if any</returns>
+    public InputEventItem Conflicts(InputEventItem item)
```
Would it be nice to get all the conflicting inputs?

FuJa0815 replied:
If there were two conflicts two separate popups would show up.

hhyyrylainen replied:
How does that look? Do you mean that the popups would overlap each other and you can only see one? Or does it actually work correctly in that all the conflicts are visible?

FuJa0815 replied:
I mean that the game should ask you two seperate time in serial.
It notifies you about the first conflict. If the user wants to remove the conflicting input the second popup should appear.
I will rewrite the code to only show one popup

hhyyrylainen commented on src/engine/input/key-mapping/InputGroupList.cs:
```diff
@@ -0,0 +1,180 @@
+    {
+        foreach (Node child in GetChildren())
+        {
+            RemoveChild(child);
```
Not 100% but I think RemoveChild doesn't actually delete the child object, so this creates a bunch of dangling objects.

FuJa0815 replied:
I have looked at other calls to RemoveChild and found out about the Free method. Ill use it here

hhyyrylainen commented on src/engine/input/key-mapping/InputGroupList.cs:
```diff
@@ -0,0 +1,180 @@
+            RemoveChild(child);
+        }
+
+        var inputGroupItemScene = GD.Load<PackedScene>("res://src/engine/input/key-mapping/InputGroupItem.tscn");
```
These scenes should be loaded just once in _Ready

hhyyrylainen commented on src/engine/input/key-mapping/InputGroupList.cs:
```diff
@@ -0,0 +1,180 @@
+        var inputActionItemScene = GD.Load<PackedScene>("res://src/engine/input/key-mapping/InputActionItem.tscn");
+
+        // Load json
+        var definition =
+            new[]
+            {
+                new
+                {
+                    EnvId = new[]
+                    {
+                        string.Empty,
+                    },
+                    GroupName = string.Empty,
+                    Actions = new[]
+                    {
+                        new
+                        {
+                            InputName = string.Empty,
+                            DisplayName = string.Empty,
+                        },
+                    },
+                },
+            };
```
This should be made into a class.

hhyyrylainen commented on src/engine/input/key-mapping/InputGroupList.cs:
```diff
@@ -0,0 +1,180 @@
+                },
+            };
+
+        using var file = new File();
```
How does this using statement not have any parenthesis or braces?

FuJa0815 replied:
its an C# 8 thing. The using without braces means that the files gets closed when the current scope (the method) get left

hhyyrylainen replied:
How does it work when we have C# target version set to 7.3?
I'm not sure if it is safe to accept this code yet? Or maybe that one guy just had outdated things? You develop on Windows, right?

FuJa0815 replied:
both windows and linux

hhyyrylainen replied:
I suppose it is fine, then. Could you edit the csproj file to specify C# 8.0, then? If someone has build issues I can always then direct them to updating their C# build tools.

hhyyrylainen replied:
Actually never mind, looks like the lang version is already set to 8.0, so everything's fine. No need for you to do anything.
I'll make a note that maybe we should port other code over to also use the C# 8 using syntax.

hhyyrylainen commented on src/engine/input/key-mapping/InputGroupList.cs:
```diff
@@ -0,0 +1,180 @@
+            };
+
+        using var file = new File();
+        file.Open("res://simulation_parameters/common/input_options.json", File.ModeFlags.Read);
```
The file path should be put in Constants

hhyyrylainen commented on src/engine/input/key-mapping/InputGroupList.cs:
```diff
@@ -0,0 +1,180 @@
+            var input = (InputGroupItem)inputGroupItemScene.Instance();
+            input.Actions = groupItem.Actions.Select(p =>
+            {
+                var inputEvents = data[p.InputName];
+
+                // Create InputActionItem
+                var inputAction = (InputActionItem)inputActionItemScene.Instance();
+                inputAction.InputName = p.InputName;
+                inputAction.DisplayName = p.DisplayName;
+                inputAction.Inputs = new ObservableCollection<InputEventItem>(
+                    inputEvents.Select(x =>
+                    {
+                        var inputEvent = (InputEventItem)InputEventItemScene.Instance();
+                        inputEvent.AssociatedEvent = x;
+                        inputEvent.AssociatedAction = inputAction;
+                        return inputEvent;
+                    }));
+                inputAction.AssociatedGroup = input;
+                return inputAction;
```
I think this code should be flattened, to me it doesn't look good to have this much code in a Select.
Using LINQ isn't applicable everywhere.

icedjuro reviewed (changes requested):
There should be some way to remove inputs.

FuJa0815 commented:
> 
> 
> There should be some way to remove inputs.

You can remove them using Escape. This should be noted somewhere

hhyyrylainen commented:
I think that's confusing, as all games that I've played, have ESC as "cancel current key change", so ESC should cancel rebinding a key. There should be like an x after each key binding that deletes it.

FuJa0815 commented:
> 
> 
> It looks like you are using just a converter to tell JSON that this is a list? You can actually make ObservableCollection implement the list interface to get that to be handled automatically.

No, I cant because ObservableCollection is an System.Collections.ObjectModel class. This is not my class.
ObservableCollection is an collection with a event for notifying about changes.

hhyyrylainen commented:
> No, I cant because ObservableCollection is an System.Collections.ObjectModel class. This is not my class.
ObservableCollection is an collection with a event for notifying about changes.

But that already inherits all the way from IList and ICollection, I think it should work correctly?
Would the thrive specific converter already included in the game work? You can put the use thrive json converter property on the type to enable that, when serialize and deserialize are used through: ThriveJsonConverter

FuJa0815 commented:
> 
> 
> > No, I cant because ObservableCollection is an System.Collections.ObjectModel class. This is not my class.
> > ObservableCollection is an collection with a event for notifying about changes.
> 
> But that already inherits all the way from IList and ICollection, I think it should work correctly?
> Would the thrive specific converter already included in the game work? You can put the use thrive json converter property on the type to enable that, when serialize and deserialize are used through: ThriveJsonConverter

Ill have a look into the thrive json converter, but I have always had problems serializing an observableList and almost always restorted to a jsonConverter

FuJa0815 commented:
You can now remove inputs by right clicking on them or pressing the little x next to it

FuJa0815 commented:
You might have to delete the options json for the changes to work

hhyyrylainen commented:
Could you resolve the conflicts and build issues before I re-review this PR?

hhyyrylainen commented on simulation_parameters/common/input_options.json:
```diff
@@ -0,0 +1,148 @@
+        "Actions": [
+            {
+                "InputName": "g_move_left",
+                "DisplayName": "Move left"
```
I think it would make translations easier if these were named "Name" to be consistent how the name is used in the other json files.

FuJa0815 replied:
Should I rename the InputName to name or the DisplayName to name?

hhyyrylainen replied:
I'd say it makes more sense to make the translation be based on DisplayName

FuJa0815 replied:
![grafik](https://user-images.githubusercontent.com/30809803/97805219-141a5500-1c55-11eb-8077-85348fd260bf.png)
should I still rename?

hhyyrylainen replied:
If you make that separate class that is loaded from the JSON there's not going to be a problem with that.

hhyyrylainen commented on src/engine/Settings.cs:
```diff
@@ -175,6 +176,14 @@ private Settings()
 
+    /// <summary>
+    ///   The current controls of the game.
+    ///   Key = InputMap name.
+    ///   Value = The events associated with the control.
```
Could you explain this a bit more elaborately.
Also this should explicitly say `key events` or input events, as just events is a bit confusing to me as it could be something else than the configured input keys.

hhyyrylainen commented on src/engine/Settings.cs:
```diff
@@ -355,6 +365,29 @@ public void ApplySoundSettings()
+    /// </summary>
+    public void ApplyInputSettings()
+    {
+        var data = CurrentControls.Value;
```
I think `data` variable could be inlined on the next line

hhyyrylainen commented on src/engine/input/key_mapping/InputActionItem.cs:
```diff
@@ -0,0 +1,141 @@
+///   Defines one specific input.
+///   Each InputActionItem has <see cref="InputEventItem">InputEventItems</see> associated with it.
+/// </summary>
+[JsonConverter(typeof(InputActionItemConverter))]
```
Could you explain where, and why you are using serialization on these GUI nodes?
This feels wrong to me. The GUI state should be easy enough to construct from the key data, right?

FuJa0815 replied:
The loading begins in this setter: https://github.com/Revolutionary-Games/Thrive/blob/9ebde8868546a8b9e8c51ed6a0848ba3f03a3c6d/src/engine/input/key_mapping/InputGroupList.cs#L51-L72
The setter is called from the InitFromData method: https://github.com/Revolutionary-Games/Thrive/blob/9ebde8868546a8b9e8c51ed6a0848ba3f03a3c6d/src/engine/input/key_mapping/InputGroupList.cs#L184-L198

It deserializes the input_options.json and stores it in a InputGroupItem List.
InputGroupItem has an JsonConverter associated with it to produce a godot ready instance of the InputGroupItem class.
InputGroupItem has those serializable properties:
- GroupName
- EnvironmentId
- Actions
Actions is an InputActionItem list and therefore triggers the InputActionItemConverter which produces a godot ready instance InputActionItem.
InputActionItem has those serializable properties:
- InputName
- DisplayName
- Inputs
Inputs is and ObservableCollection of InputEventItems.
Inputs and AssociatedGroup are set in the InputGroupItemConverter.
The InputEvent does not have an associated Converter, but gets created in the InputGroupItemConverter.

Somehow I have to load input_options.json and convert the data to godot objects. The simplest way I could imagine were JsonConverters

hhyyrylainen replied:
I get how it works now. So here's a suggestion how to do it right:
Make new classes to represent the data in the settings JSON, then make a method for creating the necessary GUI controls for the data.
This can be further split up so that the child objects of the top level GUI can also have a method "ApplyFrom(object loaded from json)" so each level of objects delegates the construction of the GUI state to their children as well.

FuJa0815 replied:
hmm, I have notices that having those changes lead to a big issue: I need to update the InputEvents and cannot create them one time.
I'm kinda having problems with created a class structure...

hhyyrylainen replied:
Here's a very rough sketch based on the JSON:

public class NamedInputAction{
public string InputName;
public string Name;
}

public class InputGroup{
public List<string> EnvironmentId;
public string GroupName;
public List<NamedInputAction> Actions;
}

Then you deserialize List<InputGroup>.
That should work.

If you want to get fancy you could make EnvironmentId into an enum so the json parse would validate that there are valid values in there.

FuJa0815 replied:
wouldn't it make more sense to make those structs?

hhyyrylainen replied:
No. structs are value types in C# meaning that they aren't reference counted and the entire objects are copied around whenever they are passed to functions etc. I wouldn't use a struct unless there were very specific requirements on the code in C#.

FuJa0815 replied:
yeah, but structs also go to the stack instead of the heap.
But you're right, using classes I can mess up less :P

FuJa0815 commented:
I hope thats what you meant by having seperate classes for the json data

hhyyrylainen commented:
If I looked at the correct commit, the new classes looked to be in the right direction.

hhyyrylainen commented on src/engine/input/key_mapping/serialisation/NamedInputGroup.cs:
```diff
@@ -0,0 +1,20 @@
+    public IReadOnlyList<string> EnvironmentId { get; set; }
+    public IReadOnlyList<NamedInputAction> Actions { get; set; }
+
+    public InputGroupItem ToGodotObject(InputDataList data)
```
So this builds an object out of this? This way of building objects isn't used as such for the other stuff.
I think it would be better if the root object of the options change hierarchy created its children, instead of doing stuff like this. Because if someone calls ToGodotObject on this and doesn't do anything with the result, that will leak the godot Node.

hhyyrylainen commented on src/engine/input/key_mapping/serialisation/NamedInputAction.cs:
```diff
@@ -0,0 +1,27 @@
+    public string InputName { get; set; }
+    public string Name { get; set; }
+
+    public InputActionItem ToGodotObject(IEnumerable<ThriveInputEventWithModifiers> data, InputGroupItem caller)
```
same thing here

hhyyrylainen commented on src/engine/input/key_mapping/ThriveInputEventWithModifiers.cs:
```diff
@@ -0,0 +1,134 @@
+using System;
+using Godot;
+
+public class ThriveInputEventWithModifiers : ICloneable
```
I had a really good name suggestion for this, but I can't remember that now. This is a bit long of a name.
Something like `InputKey` or `SpecifiedInputKey` would be better.

hhyyrylainen commented on src/engine/input/key_mapping/ThriveInputEventWithModifiers.cs:
```diff
@@ -0,0 +1,134 @@
+    /// <summary>
+    ///   Creates a string for the button to show.
+    /// </summary>
+    /// <returns>
+    ///   A human readable string.
+    /// </returns>
```
put this on a single line

hhyyrylainen commented on src/engine/input/key_mapping/ThriveInputEventWithModifiers.cs:
```diff
@@ -0,0 +1,134 @@
+        return text;
+    }
+
+    public InputEventWithModifiers ToGodotObject()
```
I think it would be better to name this more specifically like `ToInputEvent`

hhyyrylainen commented on src/engine/input/key_mapping/ThriveInputEventWithModifiers.cs:
```diff
@@ -0,0 +1,134 @@
+
+    public InputEventWithModifiers ToGodotObject()
+    {
+        InputEventWithModifiers res = Type switch
```
I think `var` would work here. Also spell out `result`

FuJa0815 replied:
var does not work here, because I have to specify the common type of the switch results

hhyyrylainen replied:
Ah, okay. I'm still not experienced enough with C# it seems, almost every time I review your PRs I come across some new syntax.

hhyyrylainen commented on src/engine/input/key_mapping/ThriveInputEventWithModifiers.cs:
```diff
@@ -0,0 +1,134 @@
+            InputType.MouseButton => new InputEventMouseButton { ButtonIndex = (int)Code },
+            _ => throw new NotSupportedException("Unsupported InputType given"),
+        };
+        res.Alt = Alt;
```
add a blank line before this.

hhyyrylainen commented on src/engine/input/key_mapping/ThriveInputEventWithModifiers.cs:
```diff
@@ -0,0 +1,134 @@
+            Alt = Alt,
+            Code = Code,
+            Control = Control,
+            Shift = Shift,
```
Does this not need to copy Type as well?

FuJa0815 replied:
forgot type and somehow worked without it :D

hhyyrylainen commented on src/engine/input/key_mapping/InputDataList.cs:
```diff
@@ -0,0 +1,32 @@
+using System.Collections.Generic;
+
+/// <summary>
+///   Stores an godot input action and its associated events
```
Doesn't this store multiple actions?

hhyyrylainen commented on src/engine/input/key_mapping/InputActionItem.cs:
```diff
@@ -0,0 +1,139 @@
+using Godot;
+
+/// <summary>
+///   Defines one specific input.
```
Isn't this for showing or editing?

FuJa0815 replied:
what do you mean?

hhyyrylainen replied:
I mean does this *literally* define one specific input. Or is this a Control for showing / editing a key input?

FuJa0815 replied:
InputActionItem itself does not handle input key editing, but it has InputEventItems as children, which handle input key editing

hhyyrylainen replied:
Could you clarify that in the doc. My goal is that if you just read the summaries of all the new classes, you get some kind of picture where abouts different input changing stuff happens.

hhyyrylainen commented on src/engine/input/key_mapping/InputActionItem.cs:
```diff
@@ -0,0 +1,139 @@
+        Inputs.CollectionChanged += OnInputsChanged;
+    }
+
+    public override bool Equals(object obj)
```
Are a bunch of these equals methods used? I'm just making sure you aren't adding these unnecessarily.

FuJa0815 replied:
They are used here: https://github.com/Revolutionary-Games/Thrive/blob/27b4b22ce91c02f48c11ab7b581039781479ead1/src/engine/input/key_mapping/InputGroupList.cs#L157-L171

hhyyrylainen replied:
Got it.

hhyyrylainen commented on src/engine/input/key_mapping/InputActionItem.cs:
```diff
@@ -0,0 +1,139 @@
+        var newInput = (InputEventItem)InputGroupList.InputEventItemScene.Instance();
+        newInput.AssociatedAction = this;
+        newInput.JustAdded = true;
+        Inputs.Add(newInput);
```
Shouldn't newInput be added as a child here?

hhyyrylainen replied:
Or does `OnInputsChanged` handle it?

FuJa0815 replied:
This happens here:
https://github.com/Revolutionary-Games/Thrive/blob/27b4b22ce91c02f48c11ab7b581039781479ead1/src/engine/input/key_mapping/InputActionItem.cs#L116-L125

hhyyrylainen commented on src/engine/input/key_mapping/InputDataList.cs:
```diff
@@ -0,0 +1,32 @@
+/// <summary>
+///   Stores an godot input action and its associated events
+/// </summary>
+public class InputDataList : ICloneable
```
This class looks pretty fine, but could the apply code from Settings be moved here, as that kinda looks into this object's privates so it would be nicer to move that into a method here and just call that from the settings.

hhyyrylainen commented on src/engine/input/key_mapping/InputEventItem.cs:
```diff
@@ -0,0 +1,210 @@
+    public bool WaitingForInput { get; private set; }
+
+    /// <summary>
+    ///   The currently assigned godot input event.
```
isn't `godot input event` wrong?

hhyyrylainen commented on src/engine/input/key_mapping/InputEventItem.cs:
```diff
@@ -0,0 +1,210 @@
+    internal bool JustAdded { get; set; }
+
+    /// <summary>
+    ///   Deleted this event from the associated action and updated the godot InputMap
```
`Delete`

hhyyrylainen commented on src/engine/input/key_mapping/InputEventItem.cs:
```diff
@@ -0,0 +1,210 @@
+    public void Delete()
+    {
+        AssociatedAction.Inputs.Remove(this);
+        InputGroupList.Instance.ControlsChanged();
```
Instead of accessing a global, would it work if that was passed in as a property through the node tree until it reaches here?

FuJa0815 replied:
I honestly don't get what you mean...

hhyyrylainen replied:
I mean that, you create the tree of objects starting from the input group list, which in turn ends up creating this.
Global use should be avoided.
So I suggest you do something like:
input group list -> creates some object and passes itself as reference -> this in turn creates then InputEventItem and passes the reference to input group list

That way there isn't a need to use a global here. As I said, if feasible in cases like this instead of using a global the object should be propagated through to where it is needed.

FuJa0815 replied:
with my latest commits I have made many things static. Did this resolve itself or should I still change this?

hhyyrylainen replied:
See my latest comment: https://github.com/Revolutionary-Games/Thrive/pull/1772#discussion_r516594823
It also applies here

hhyyrylainen commented on src/engine/input/key_mapping/InputEventItem.cs:
```diff
@@ -0,0 +1,210 @@
+
+    protected override void Dispose(bool disposing)
+    {
+        AssociatedAction = null;
+        button?.Dispose();
+        xbutton?.Dispose();
+        base.Dispose(disposing);
```
This doesn't seem to dispose any classes that want to be disposed, so can't this be removed?

FuJa0815 replied:
the important thing in the dispose method is the 
`AssociatedAction = null;`
line.
Everything else is to satisfy CI

FuJa0815 replied:
Setting AssociatedAction to null is important for the GC because otherwise I would have an unhandled reference loop

hhyyrylainen replied:
As long as you don't reference it externally, it should get deleted? Isn't C# smart enough that it handles reference loops that aren't referenced from outside? Or you could put in a weak reference into the AssociatedAction or make AssociatedAction a weak reference.

FuJa0815 replied:
C#'s GC odd being smart enough, but I've experienced that loops can still cause memory problems. Just fixing this in the Dispose is the easiest way to make sure there aren't any problems

hhyyrylainen replied:
Anyone doing further changes needs to be careful with this class, then.
Weak references and refactoring could be done to make this less of a landmine to step on later.

FuJa0815 replied:
Is it really a landmine? I mean, CI warns you if you forget to dispose something

hhyyrylainen replied:
That's true, but I'd really like to avoid the extra dispose method if at all possible.

hhyyrylainen commented on src/engine/input/key_mapping/InputGroupItem.cs:
```diff
@@ -0,0 +1,66 @@
+    /// </example>
+    public string GroupName { get; set; }
+
+    // ReSharper disable once CollectionNeverUpdated.Global (Reason: Loaded from json)
```
Leftover comment?

hhyyrylainen reviewed (changes requested):
Didn't get through everything but here are some more comments.

FuJa0815 commented on src/general/OptionsMenu.cs:
```diff
@@ -952,6 +994,11 @@ private void OnTutorialsEnabledToggled(bool pressed)
 
+    private void InitInputGroupList()
+    {
+        InputGroupList.Instance.InitFromData(Settings.Instance.CurrentControls);
```
but I cant make methods like ShowInputConflictDialog static. How should I call this method from InputEventItem?

FuJa0815 replied:
I hope my commit is what you mean

hhyyrylainen replied:
Sorry for not being able to reply earlier, I had some other stuff to do.
You didn't do what I had in mind, it's a bit better to have static methods than a global object, but I feel like you just basically converted the global object to a static class. So it isn't much better.

Even if there is only ever one of the input group things at once, I think it's good to ask "should the input list access the global group object *or* should it access the group object that owns it?" To me it is obvious that the second one is right.

So this is the kind of class structure I recommend:

```
class InputGroupThing{
public void BuildGUI(){
var child = blah.Instance();
child.GroupOwner = this;
child.BuildGUI();
}
}

class FirstLevelChild{
public InputGroupThing {get; set;}

public void BuildGUI(){
var anotherChild = bleh.Instance();
anotherChild.GroupOwner = InputGroupThing;
anotherChild.BuildGUI();
}
}

```
That's what I meant when I said to flow / pass the input group reference through the object tree.

FuJa0815 replied:
Ok, I think I get what you mean.
Please check my commit and check if I did finally understand what you meant :D

hhyyrylainen replied:
You created the GUI Control tree like I suggested, but I don't see you removing the pseudo-global usage?

FuJa0815 replied:
I have to do some pseudo global access, but I reduced it drasticly. Is this ok now?

hhyyrylainen replied:
It's better now.

Related to the new code, I'm not sure I agree with how much null checks there are and ? usage. Shouldn't the GUI for key changing be always in a valid state? Having those ignore if null, could make further development much harder, if they mask null problems after refactoring or something.

FuJa0815 replied:
There shouldn't be any nulls, but if the weak reference was disposed, there would be null pointers everywhere otherwise. I think its safer to check if the weak reference is disposed (and my methods therefore returns null) instead of just blindly relying on it not being disposed.

hhyyrylainen replied:
I think it should *never* be the case that the children of an input group are destroyed *after* the group. So the problem condition should never happen. For that reason I kinda would like to have the code just go ahead and flat out crash with null pointer or disposed exception, as when refactoring or doing other changes to the input binding, it will be faster to get exceptions when you do something wrong.
@Kasterisk might be able to weigh in on this.

Kasterisk replied:
I'm afraid I wouldn't be much help with the workings of the code, but I'll just go with hhyyrylainen on this.

Kasterisk reviewed (commented):
I think currently you could only remove a keybind while adding and selecting a key for it, shouldn't it also be possible to remove new keybind after it's added? Also maybe there could be a reset inputs button somewhere IMO.

FuJa0815 commented:
@Kasterisk You can remove a bind once its added by either leftclicking on it and pressing the X button or by simply right clicking it.
There is no seperate reset inputs button, but the **Defaults** and **Reset** buttons work.

hhyyrylainen commented:
I think it is a bit useful to be able to reset just the controls, and not volume and window settings for example.

FuJa0815 commented:
Umm, have I done something wrong with the format_script?

Btw, the Translations:
- INPUTS (the Options tab text)
- RESET_INPUTS (the button text)
- RESET_INPUTS_TO_DEFAULTS (popup title)
- ARE_YOU_SURE_TO_RESET_INPUT_SETTINGS (popup content)
are new.

hhyyrylainen commented:
> Umm, have I done something wrong with the format_script?

How so?

FuJa0815 commented:
Whops, my fault, should have read the new working with translations more closely

hhyyrylainen commented on project.godot:
```diff
@@ -40,6 +40,7 @@ ToolTipManager="*res://src/gui_common/tooltip/ToolTipManager.tscn"
 TransitionManager="*res://src/gui_common/TransitionManager.cs"
 CursorLoader="*res://src/engine/CursorLoader.cs"
 ScreenShotTaker="*res://src/engine/ScreenShotTaker.cs"
+QuickLoadHandler="*res://src/gui_common/QuickLoadHandler.cs"
```
This is pretty strange. I wonder where this change is coming from?

hhyyrylainen replied:
Also it seems the game doesn't currently start. I'm all out of energy for today so I'm going to check out things tomorrow (if someone hasn't fixed things by then).

FuJa0815 replied:
I've removed it and fixed a few external assignments and now it is working

hhyyrylainen commented:
Btw, I updated the required Babel-Thrive version. That's probably why the translations check is still failing.
I can quickly fix that and re-squash the last few commits tomorrow before merging.

hhyyrylainen commented:
Actually I'll do that now to give people some time to look at the final code to be merged tomorrow.

hhyyrylainen reviewed (approved):
Did the changes I want. I'll open a couple of follow up issues once I merge this.

Kasterisk commented on simulation_parameters/SimulationParameters.cs:
```diff
@@ -62,6 +63,8 @@ private SimulationParameters()
 
         helpTexts = LoadRegistry<HelpTexts>("res://simulation_parameters/common/help_texts.json");
 
+        inputGroups = LoadListRegistry<NamedInputGroup>(Constants.INPUT_OPTIONS);
```
I'm guessing the other registry loads in `SimulationParameters` will be changed to also use constants for the path. If not, shouldn't this pass the string of the full path directly? (since the other loads do that)

hhyyrylainen replied:
I just copied over the use of the constant. It's kinda redundant here as only the simulation paramters should ever need to know the path to the JSON files it loads. I just couldn't be bothered to think and change it as I'd already spent so much time on this PR.

hhyyrylainen replied:
Inlined this string.

Kasterisk reviewed (changes requested):
There are some UI stuff for the Inputs tab that I would like to tweak and improve a bit, I'll push my changes here if that's fine.

hhyyrylainen commented:
It's fine to put your changes here, just try to not make "dirty" (small CI fixes, a couple line tweaks after a big commit) commits as I've cleaned this branch

hhyyrylainen commented on src/engine/input/key_mapping/InputActionItem.cs:
```diff
@@ -0,0 +1,177 @@
+    }
+
+    /// <summary>
+    ///   Sets up the inputs and adds them as children.
```
Isn't "adds" grammatically correct here?

Kasterisk replied:
It seems so.

hhyyrylainen commented on src/engine/input/key_mapping/InputEventItem.cs:
```diff
@@ -0,0 +1,293 @@
+                    continue;
+
+                // Set AssociatedEvent to null to not delete the wrong InputEventItem,
+                // because Equals threats it the same with the same AssociatedEvent.
```
Probably should be `treats`

hhyyrylainen commented on src/engine/input/key_mapping/InputGroupList.cs:
```diff
@@ -0,0 +1,230 @@
+    {
+        foreach (Node child in GetChildren())
+        {
+            RemoveChild(child);
+
+            child.QueueFree();
```
Changed this to just Free()

hhyyrylainen commented:
Btw @Kasterisk I noticed on your last commit that you updated a bunch of the translation files, but they had the wrong line numbers, for example I just now reran the script and I got this:
```diff
-#: ../src/auto-evo/AutoEvoRun.cs:158
+#: ../src/auto-evo/AutoEvoRun.cs:119
 msgid "ABORTED"
 msgstr "Прекратена."
```
Checking the file I was able to confirm that string is actually on line 119.

hhyyrylainen commented:
Actually now I get a disposed error when I removed that remove from parent call. And the problem went away.
Could this be an actual problem? Could the saving issues also be caused by this? This is a huge finding, I think. I'll open an issue later to review all places where Free or QueueFree is used to see if remove from parent calls should also be added there.

hhyyrylainen commented:
I'm going to merge this once CI checks pass.

RacerBG commented:
@hhyyrylainen I see that some of the new translatable strings have messed up translation. This does not affect the already translated strings, is that correct?

Thanksfully, the new ones which already have incorrect translation are marked as "Needs editing".

hhyyrylainen commented:
I did notice that as well. It seems that when new strings to be translated are added, gettext tries to find already translated things that are close enough and fill those in. Which might seem a bit counter-intuitive, but I let it pass because it correctly marks those "guessed" translations as needing editing.

RacerBG commented:
> 
> 
> I did notice that as well. It seems that when new strings to be translated are added, gettext tries to find already translated things that are close enough and fill those in. Which might seem a bit counter-intuitive, but I let it pass because it correctly marks those "guessed" translations as needing editing.

Thanks for clarifying that for me.
